stor/obs.py

```diff
@@ -303,6 +326,7 @@ def __init__(self, pth, mode='r', encoding=None, **kwargs):
         self.mode = mode
         self._kwargs = kwargs
         self.encoding = encoding or locale.getpreferredencoding(False)
+        self._path.encoding = self.encoding
```

Comment:
rather than have magic encoding method, let's update API for read_object to support passing in an encoding, with fall back of `locale.getpreferredencoding(False)` .

This also suggests you need a test case for calling `read_object()` on a DXPath that was not accessed via open(), because currently that would fail

Reply:
Do you want me to update this api only for DXPath, or other OBSPath too (which would ignore the encoding argument)?

Reply:
Actually, after looking into dxpy3, DXFile.read method decodes the read data using 'utf-8' directly if it is Python3, which means that if we write a 'utf-16' file to DNAnexus (using stor), we cannot read it using dxpy3. Hence, removing the encoding argument, and hardcoding it with a note for now.

Reply:
The API has to be cross-compatible, thus all implementations need to support the same keyword arguments. Currently we support `encoding` argument for Swift and S3, thus we need it here.

My suggestion would be, for DNAnexus, to check if the mode is `w` or `r` and assert that the encoding is `utf8` (e.g.

```
if encoding and encoding not in ('utf-8', 'utf8'):
    raise ValueError('For DNAnexus paths, text encoding is always assumed to be utf8. Open in binary mode and decode manually to use an alternate encoding.')
```

Reply:
I meant I hardcoded the encoding in `read_object`. open still has an `encoding` argument of course. `read_object` doesn't have an encoding argument in swift/s3/dx now.

Also, 'rb' won't help with reading other objects in dxpy3 either, referenced issue elsewhere. I can do something like
```
if six.PY3 and encoding and encoding not in ('utf-8', 'utf8'):
    raise ValueError('For DNAnexus paths in Python 3, encoding is always assumed to be 'utf-8'. Please switch your encoding or Python version')
```
Currently the dxpy error bubbles up. I was thinking of just waiting for them to fix this issue before addressing this, but not knowing when that'll be, adding this error message now.